### ISSUE TITLE
DOC: Subarray Iteration with nditer

### ISSUE THREAD

**Issue by OyiboRivers** [numpy/numpy#25293]
### Issue with current documentation:

Dear Numpy Team,

I am trying to use numpy.nditer to iterate over subarrays of ndarrays. While working on this, I encountered some challenges and would appreciate your insights on the best practices for such iterations.

I am referring to:
https://numpy.org/doc/stable/reference/generated/numpy.nditer.html
https://numpy.org/doc/stable/reference/arrays.nditer.html

Here are some observations and questions based on my experimentation:

To obtain subarrays on each iteration step, it is necessary to use the "external_loop" flag, otherwise, you receive element-wise values.

For column-wise iteration over a C-contiguous array, setting the iteration order to "F" (Fortran) was required to avoid a flattened array. The result needed transposition to return to a C-contiguous array.

For row-wise iteration over a C-contiguous array, I initially copied and transposed the input array, and again set the iteration order to "F." The result required transposition to return to a C-contiguous array. Can the copy process be avoided?

Handling multiple ndarrays in such iterations could potentially become complex, and I suspect there might be more straightforward ways to achieve the same results.

I have included a code example to illustrate the challenges and approaches I've taken. It would be very helpful if you could share any best practices or recommendations for efficiently iterating over subarrays using numpy.nditer in Python and/or Cython.

```
import numpy as np

def iter_array(arr: np.ndarray,
               axis: int | None = None,
               res: np.ndarray | None = None) -> np.ndarray:
    """
    Function iterates over subarrays of an ndarray based on the specified axis.

    Parameters:
    - arr (np.ndarray): The input array to be iterated over.
    - axis (int, optional): The axis along which to iterate.
        If None, the array is flattened.
    - res (np.ndarray, optional): An array to store the result of the iteration.
        If None, a new array is allocated.

    Returns:
    - np.ndarray: The result array after iteration.

    Example:
        arr = np.arange(6.).reshape(2, 3)
        iter_array(arr, axis=None)
        iter_array(arr, axis=0)
        iter_array(arr, axis=1)
    """
    # ***** Print some information *****
    msg = 'iterate C_CONTIGUOUS array '
    if axis == 0:
        msg += ('column-wise' + '\n'
                + '(order=F, transpose output)')
    elif axis == 1:
        msg += ('row-wise' + '\n'
                + '(order=F, tanspose+copy input, transpose output)')
    else:
        msg += 'flattened' + '\n' + '(order=k)'
    print('=============================================')
    print(msg)
    print('=============================================')
    print('Input array:', '\n', arr, '\n')

    # ***** Iteration process *****
    a = arr.T.copy() if axis == 1 else arr
    order = 'k' if axis is None else 'F'
    with np.nditer(
        op=[a, res],
        flags=['external_loop'],
        op_flags=[['readonly'],                 # Input flags
                  ['writeonly', 'allocate']],   # Output flags
        order=order,
    ) as it:
        for it_in, it_out in it:
            it_out[...] = it_in - np.mean(it_in)
            print(f'iterator: {it_in} do some math => x-xmean = {it_out}')
        res = it.operands[1] if axis is None else it.operands[1].T
    print('\nResult array:', '\n', res, '\n')
    return res

# =============================================
# iterate C_CONTIGUOUS array flattened
# (order=k)
# =============================================
# Input array:
#  [[0. 1. 2.]
#  [3. 4. 5.]]

# iterator: [0. 1. 2. 3. 4. 5.] do some math => x-xmean = [-2.5 -1.5 -0.5  0.5  1.5  2.5]

# Result array:
#  [[-2.5 -1.5 -0.5]
#  [ 0.5  1.5  2.5]]

# =============================================
# iterate C_CONTIGUOUS array column-wise
# (order=F, transpose output)
# =============================================
# Input array:
#  [[0. 1. 2.]
#  [3. 4. 5.]]

# iterator: [0. 3.] do some math => x-xmean = [-1.5  1.5]
# iterator: [1. 4.] do some math => x-xmean = [-1.5  1.5]
# iterator: [2. 5.] do some math => x-xmean = [-1.5  1.5]

# Result array:
#  [[-1.5  1.5]
#  [-1.5  1.5]
#  [-1.5  1.5]]

# =============================================
# iterate C_CONTIGUOUS array row-wise
# (order=F, tanspose+copy input, transpose output)
# =============================================
# Input array:
#  [[0. 1. 2.]
#  [3. 4. 5.]]

# iterator: [0. 1. 2.] do some math => x-xmean = [-1.  0.  1.]
# iterator: [3. 4. 5.] do some math => x-xmean = [-1.  0.  1.]

# Result array:
#  [[-1.  0.  1.]
#  [-1.  0.  1.]]
```

### Idea or request for content:

Idea/Request: Extending "Iterating Over Arrays" Documentation
https://numpy.org/doc/stable/reference/arrays.nditer.html

Add a subsection on using nditer for iterating over subarrays.
Provide guidelines and code examples of using nditer in different scenarios i.e operating on multiple arrays in different dimensions.
Extend the section on array iteration to include performance considerations, highlighting potential bottlenecks and optimizations.
How does the performance of nditer compare to Python or C iterations.
Include examples of i.e. Cython memory views and buffer access for efficient array iteration, along with strategies for minimizing Python overhead.

Regards Oyibo

**Comment by seberg** [comment]
`nditer` is really a very raw API and more useful to understand the C-part in many ways.  It isn't meant to do what you want to do.  One can remove axes to exclude them from iteration, but again this is going to be very raw/unsafe in Python.

There is `nested_nditer` or so, which might help.  Overall, I would suggest to just use `np.ndindex()` or a direct iteration and using normal indexing instead of `nditer`.

**Comment by OyiboRivers** [comment]
Hello @seberg ,

Thank you very much for your quick response regarding `nditer`.

I have a follow-up question to make sure I got it right. In your message, you mentioned that nditer is more of a raw API useful for understanding the C-part. I was wondering if this advice is specific to subarray iteration or if it extends to general array iteration in Python. Should nditer be avoided where other iteration methods are available in Python (and C)?

`nested_nditer` does not exist. You probably mean `nested_iters` which returns a tuple of `nditer`.

On a related note is there a complete code example available how to define a custom loop function for generalized universal functions for the C-API `PyUFunc_FromFuncAndDataAndSignature` using Cython? I was exloring if nditer could be useful here.

Thanks again for your time and effort.

**Comment by OyiboRivers** [comment]
Hello @seeberg,

Thank you once again for sharing your thoughts.

I believe there might be a slight miscommunication in our discussion, and I'd like to clarify a bit.

My goal is not to create universal functions (ufuncs) but rather generalized universal functions (gufuncs) or functions with similar capabilities. Specifically, I aim to implement these in Cython to generate C-code.

In the NumPy documentation under "Iterating over arrays," there's an example using nditer for efficient array iteration. I experimented with nditer in Python to understand how it works and how to tune the parameters. I checked the possibility of utilizing the iterator to iterate not only element-wise but also over entire columns and rows by adjusting the memory order.

As you pointed out already, nditer or NpyIter might not be the optimal choices for this use case.

Additionally, I'm uncertain whether using nditer in Cython provides access to the C-interface. This is something I could find out from the Cython team. Nevertheless, I could use NpyIter instead if it would be more suitable.

To avoid any confusion, let me clarify some definitions:

Ufuncs perform element-wise operations with scalar arguments.
```
@ufunc(signature="(),()->()")
def foo(a: float, b: float) -> float:
    return a + b
```

Gufuncs extend ufuncs to support more complex operations and custom broadcasting rules. The core dimensions can be scalars or ndarrays.
```
@gufunc(signature="(k?),(m?,n)->(k?,m?)")
def npv(rate: float, values: np.ndarray[np.float64]) -> float:
    res = 0.0
    for timestep in range(values.shape[0]):
        res += values[timestep] / ((1.0 + rate) ** timestep)
    return res
```
While Cython provides a decorator for generating ufuncs, there is currently no tool for generating custom gufuncs. Numba, on the other hand, offers a "vectorize" decorator for ufuncs and a "guvectorize" decorator for gufuncs. However, there are limitations, such as the inability to use these functions in other jitted functions and the lack of support for NEP 20 signatures like "(k?),(m?,n)->(k?,m?)".

I haven't found a tool to assist in generating custom gufuncs. This led me to consider using Cython and the NumPy C-Interface, but I faced problems in defining the first two arguments of PyUFunc_FromFuncAndDataAndSignature (PyUFuncGenericFunction *func, void *const *data, ...) correctly. Tutorials with comprehensive examples for this process are not easy to find.

In conclusion, the alternative approach is to write functions entirely in Cython or C and explicitly implement the iteration logic. However, even a simple function signature like 
`(k?),(m?,n)->(k?,m?)`
 generates multiple variations, such as 
```
['(),(n)->()',
 '(),(m,n)->(m)',
 '(k),(n)->(k)',
 '(k),(m,n)->(k,m)']
```
You can easily miss cases or mix something up. That is a disadvantage compared to an automated process. At the end you want to write less amount of code while ensuring safety and performance.

**Comment by seberg** [comment]
OK, so you want C/Cython and not `nditer`, so you are using `nditer` in the way I think it is mostly useful (unless you know it already, there are some Python use-cases, but all of them come with caveats pretty much).

For GUFuncs, I dunno.  TBH, while I wouldn't mind having Cython docs, I also don't think it is unreasonable to only have C ones.  Now we probably don't have C ones either really (maybe the NEP has some info), but you can find some being generated in our tests for example.
Another thing is that we have some new API upcoming, so that makes me want to look into it even less.

Having problems defining the first two arguments seems just like two small Cython hacks to figure out?
Cython might be annoying around `const`s?! What is your *actual* problem, you still didn't really say.  Maybe just define everything with `void *` in Cython and cast to that when passing it...

For `NpyIter` I still do *not* think you want to use it, but we just added the `NpyIter` API into the cython `__init__.pyd` file, you could copy it out if you wish to use it in Cython.  But, it is a very C-like API, so you need to take a lot of care with cleanup, etc.

The important point there, as I said, is `op_axes` which selects all but the core dimensions, using that from Python gives you a single element I think, but that is a pointer to the start of the subarray you want.
That is what NumPy uses internally to implement gufuncs, although casts, etc. are more clunky for them (and could be improved).

**Comment by OyiboRivers** [comment]
Hello @seberg ,

thank you very much.
I have found some resources including the one you shared that should help me to get over the hill.

https://scipy-lectures.org/advanced/advanced_numpy/#generalized-ufuncs
https://github.com/xnd-project/gufunctools/blob/master/modules/examples/src/example_gufuncs.c
https://github.com/numpy/numpy/blob/main/numpy/_core/src/umath/_umath_tests.c.src
https://github.com/WarrenWeckesser/ufunclab/tree/main

> Another thing is that we have some new API upcoming, so that makes me want to look into it even less.

Regarding the upcoming API changes, I understand your perspective on waiting for the new API.
I'll keep an eye out for the release of Numpy 2.0 in the coming weeks/months.

Once again, thank you for your help.
